File: packages/kits/AriaMcpKit/Sources/AriaMCP/MonitoringControl.swift
// MonitoringControl.swift
//
// Injection seam for daemon monitoring state. AriaMcpKit defines the protocol;
// the serve host layer (AriaResident in Swift, runtime.rs in Rust) provides
// the concrete implementation wrapping the StatsStore. This keeps AriaMcpKit
// free of ObserverSink / IntellectusLib — the separation established in the
// resident-mode architecture.
//
// The read-write split maps to the two call paths of `moot_monitoring_status`:
//   - absent `enabled` arg  → read()  → reports current effective state
//   - present `enabled` arg → set(_:) → writes flag, reports new effective state
//
// Nil from read() means the store is present but the flag cannot be read
// (transient error). Return nil from a nil-monitoringControl means no store
// is wired at all. Both cases report "unavailable" in the tool response so
// state is never fabricated — never invent "disabled" when the true answer
// is "unknown".

import Foundation

/// Injection seam for daemon telemetry monitoring state.
///
/// The concrete implementation lives in the serve host layer (AriaResident)
/// wrapping a `StatsStore`. AriaMcpKit stores only this protocol reference —
/// no ObserverSink import, no IntellectusLib import.
///
/// - `read()` returns `nil` on transient store error or when no store is wired.
/// - `set(_:)` persists the flag; best-effort (errors are logged by the
///   implementation, not surfaced to the caller).
///
/// Both methods are `async` because `StatsStore` is an actor and its API is
/// `async throws`. The Rust mirror uses sync equivalents.
public protocol MonitoringControl: Sendable {
    /// Return the current monitoring-enabled flag, or `nil` if it cannot be
    /// read (transient store error). `nil` ≠ `false` — do not substitute one
    /// for the other in any response.
    func read() async -> Bool?

    /// Persist `enabled` as the new monitoring flag. Best-effort: errors are
    /// logged by the implementation and do not throw. Callers may follow with
    /// `read()` to confirm the persisted value when accuracy matters.
    func set(_ enabled: Bool) async
}

File: packages/kits/AriaMcpKit/rust/src/monitoring_control.rs
//! MonitoringControl — injectable monitoring gate for `moot_monitoring_status`.
//!
//! Mirrors Swift `MonitoringControl` protocol. AriaMcpKit defines the trait;
//! the serve host supplies the concrete impl that wraps `observer_sink::StatsStore`.
//! This keeps AriaMcpKit free of the observer_sink dependency — the same
//! boundary discipline used for vault tools and sensitivity ledgers.
//!
//! # Unavailable case
//!
//! When no stats store is wired (stdio mode, test harnesses, provision-less
//! contexts), the dispatcher carries `None` for this field. The tool runner
//! (`run_monitoring_status` in `interface_tools.rs`) reports
//! "monitoring: unavailable" in that case — it never fabricates a state.
//!
//! # Threading
//!
//! `MonitoringControl: Send + Sync` because it may be shared behind `Arc` across
//! threads in the HTTP server path. `StatsStoreMonitoringControl` satisfies this
//! because `observer_sink::StatsStore` is `Send + Sync`.

use std::sync::Arc;

/// Read/set the daemon's monitoring flag without importing observer_sink directly.
///
/// Methods are infallible from the caller's perspective — errors are absorbed
/// by implementations and surface as `None` on read or are silently logged on
/// write. This matches the Swift protocol's best-effort `set` and nil-returning
/// `read` contract.
pub trait MonitoringControl: Send + Sync {
    /// Return the current monitoring-enabled flag.
    /// Returns `None` when the store cannot be read (transient error).
    /// `None` ≠ `false` — callers must not substitute one for the other.
    fn read(&self) -> Option<bool>;

    /// Persist `enabled` as the new monitoring flag.
    /// Best-effort: errors are swallowed by the implementation.
    /// The store also writes `monitoring_source=user` so downstream
    /// readers can distinguish operator-driven changes from defaults.
    fn set(&self, enabled: bool);
}

/// Concrete `MonitoringControl` backed by an `observer_sink::StatsStore`.
///
/// Constructed by `run_http_loop` from the `http_stats_store` when available.
/// `None` for the store in the `Dispatcher` means this type is never constructed.
pub struct StatsStoreMonitoringControl {
    pub store: Arc<observer_sink::StatsStore>,
}

impl MonitoringControl for StatsStoreMonitoringControl {
    fn read(&self) -> Option<bool> {
        self.store.is_monitoring_enabled().ok()
    }

    fn set(&self, enabled: bool) {
        // Best-effort — log on error but do not propagate.
        if let Err(e) = self.store.set_monitoring_enabled(enabled) {
            eprintln!("[moot_monitoring_status] set_monitoring_enabled failed: {e}");
        }
    }
}
